rclrs/src/clock.rs

```diff
@@ -0,0 +1,221 @@
+    fn from(clock_type: ClockType) -> Self {
+        match clock_type {
+            ClockType::RosTime => rcl_clock_type_t::RCL_ROS_TIME,
+            ClockType::SystemTime => rcl_clock_type_t::RCL_SYSTEM_TIME,
+            ClockType::SteadyTime => rcl_clock_type_t::RCL_STEADY_TIME,
+        }
+    }
+}
+
+/// Struct that implements a Clock and wraps `rcl_clock_t`.
+#[derive(Clone, Debug)]
+pub struct Clock {
+    _type: ClockType,
+    _rcl_clock: Arc<Mutex<rcl_clock_t>>,
+    // TODO(luca) Implement jump callbacks
+}
+
+/// A clock source that can be used to drive the contained clock. Created when a clock of type
+/// `ClockType::RosTime` is constructed
+pub struct ClockSource {
+    _rcl_clock: Arc<Mutex<rcl_clock_t>>,
+}
+
+impl Clock {
+    /// Creates a new Clock with `ClockType::SystemTime`
+    pub fn system() -> Result<Self, RclrsError> {
+        Self::make(ClockType::SystemTime)
+    }
+
+    /// Creates a new Clock with `ClockType::SteadyTime`
+    pub fn steady() -> Result<Self, RclrsError> {
+        Self::make(ClockType::SteadyTime)
+    }
+
+    /// Creates a new Clock with `ClockType::RosTime` and a matching `ClockSource` that can be used
+    /// to update it
+    pub fn with_source() -> Result<(Self, ClockSource), RclrsError> {
+        let clock = Self::make(ClockType::RosTime)?;
+        let clock_source = ClockSource::new(clock._rcl_clock.clone());
+        Ok((clock, clock_source))
+    }
+
+    /// Creates a new clock of the given `ClockType`.
+    pub fn new(type_: ClockType) -> Result<(Self, Option<ClockSource>), RclrsError> {
+        let clock = Self::make(type_)?;
+        let clock_source =
+            matches!(type_, ClockType::RosTime).then(|| ClockSource::new(clock._rcl_clock.clone()));
+        Ok((clock, clock_source))
+    }
+
+    fn make(type_: ClockType) -> Result<Self, RclrsError> {
+        let mut rcl_clock;
+        unsafe {
+            // SAFETY: Getting a default value is always safe.
+            rcl_clock = Self::init_generic_clock();
+            let mut allocator = rcutils_get_default_allocator();
+            rcl_clock_init(type_.into(), &mut rcl_clock, &mut allocator).ok()?;
+        }
+        Ok(Self {
+            _type: type_,
+            _rcl_clock: Arc::new(Mutex::new(rcl_clock)),
+        })
+    }
+
+    /// Returns the clock's `ClockType`.
+    pub fn clock_type(&self) -> ClockType {
+        self._type
+    }
+
+    /// Returns the current clock's timestamp.
+    pub fn now(&self) -> Time {
+        let mut clock = self._rcl_clock.lock().unwrap();
+        let mut time_point: i64 = 0;
+        unsafe {
+            // SAFETY: No preconditions for his function
+            rcl_clock_get_now(&mut *clock, &mut time_point);
+        }
+        Time {
+            nsec: time_point,
+            clock: Arc::downgrade(&self._rcl_clock),
+        }
+    }
+
+    /// Helper function to initialize a default clock, same behavior as `rcl_init_generic_clock`.
```

Comment:
nit: Perhaps consider rewording slightly to something like this? I had a hard time following at first

> Helper function to privately initialize a default clock. This function has the same behavior as `rcl_init_generic_clock`. However, by defining a function like this, instead of implementing `Default`, we avoid exposing a public API to create an invalid clock.

I think the main point to convey is that we'd rather define this privately than in a public API

Reply:
https://github.com/ros2-rust/ros2_rust/pull/325/commits/ed583f67d44e382015766a325041b8b5389dfd38. Reworded a bit to avoid saying the word `function` three times in a row but kept most of your suggestion.